src/compiler/crystal/semantic/bindings.cr

```diff
@@ -1,7 +1,69 @@
 module Crystal
+  # Specialized container for ASTNodes to use for bindings tracking.
+  #
+  # The average number of elements in both dependencies and observers is below 2
+  # for ASTNodes. This struct inlines the first two elements saving up 4
+  # allocations per node (two arrays, with a header and buffer for each) but we
+  # need to pay a slight extra cost in memory upfront: a total of 6 pointers (48
+  # bytes) vs 2 pointers (16 bytes). The other downside is that since this is a
+  # struct, we need to be careful with mutation.
+  struct SmallNodeList
+    include Enumerable(ASTNode)
+
+    @first : ASTNode?
+    @second : ASTNode?
+    @tail : Array(ASTNode)?
+
+    def each(& : ASTNode ->)
+      if first = @first
+        yield first
+      end
+      if second = @second
+        yield second
+      end
+      if tail = @tail
+        tail.each { |node| yield node }
+      end
+    end
+
+    def push(node : ASTNode) : self
+      if @first.nil?
+        @first = node
+      elsif @second.nil?
+        @second = node
+      else
+        tail = @tail ||= Array(ASTNode).new
+        tail.push(node)
+      end
+      self
+    end
+
+    def reject!(& : ASTNode ->) : self
+      if first = @first
+        if yield first
+          @first = nil
+        end
+      end
+      if second = @second
+        if yield second
+          @second = nil
+        end
+      end
+      if tail = @tail
+        tail.reject! { |node| yield node }
+      end
+      self
+    end
```

Comment:
I converted this collection so that it never has "holes". `reject!` needs to have a much more complicated implementation (so the code is harder to trust), but in turn everything else can have an early return. Also it preserves push order.

Feel free to ignore this, I just wanted to have fun :D

```suggestion
    def each(& : ASTNode ->)
      if first = @first
        yield first
        if second = @second
          yield second
          if tail = @tail
            tail.each { |node| yield node }
          end
        end
      end
    end

    def size
      if @first.nil?
        0
      elsif @second.nil?
        1
      elsif (tail = @tail).nil?
        2
      else
        2 + tail.size
      end
    end

    def push(node : ASTNode) : self
      if @first.nil?
        @first = node
      elsif @second.nil?
        @second = node
      elsif (tail = @tail).nil?
        @tail = [node] of ASTNode
      else
        tail.push(node)
      end
      self
    end

    def reject!(& : ASTNode ->) : self
      if first = @first
        if second = @second
          if tail = @tail
            tail.reject! { |node| yield node }
          end
          if yield second
            @second = tail.try &.shift?
          end
        end
        if yield first
          @first = @second
          @second = tail.try &.shift?
        end
      end
      self
    end
```

I pushed a commit to my fork, that way it's running CI immediately so we can confirm this works. https://github.com/oprypin/crystal/commit/f967adeaa2a1fb19cfa62dbf32591592fbea31d1?diff=split

Reply:
Nice fun. I believe `#each` can actually be shortened a bit more:
```cr
def each(&)
  yield @first || return
  yield @second || return
  @tail.try(&.each { |node| yield node })
end
```

Reply:
I incorporated this change as well, thanks!